examples/core/graphsage/link_prediction.py

```diff
@@ -0,0 +1,265 @@
+
+        # Compute prediction scores using the model.
+        pred = model.predictor(h_src * h_dst).squeeze(-1)
+
+        # Evaluate the predictions to obtain MRR values.
+        input_dict = {"y_pred_pos": pred[:, 0], "y_pred_neg": pred[:, 1:]}
+        rr[start:end] = evaluator.eval(input_dict)["mrr_list"]
+    return rr.mean()
+
+
+@torch.no_grad()
+def evaluate(device, graph, edge_split, model):
+    """Evaluate the model on validation and test sets."""
+    model.eval()
+    evaluator = Evaluator(name="ogbl-citation2")
+
+    # Compute node embeddings for the entire graph.
+    node_emb = model.inference(graph)
+    results = []
+
+    # Loop over both validation and test sets.
+    for split in ["valid", "test"]:
+        src = edge_split[split]["source_node"].to(node_emb.device)
+        dst = edge_split[split]["target_node"].to(node_emb.device)
+        neg_dst = edge_split[split]["target_node_neg"].to(node_emb.device)
+
+        # Compute MRR values for the current split.
+        results.append(
+            compute_mrr(model, evaluator, node_emb, src, dst, neg_dst, device)
+        )
+    return results
+
+
+def construct_negative_graph(graph, k):
+    src, dst = graph.edges()
+
+    neg_src = src.repeat_interleave(k)
+    neg_dst = torch.randint(0, graph.num_nodes(), (len(src) * k,))
+    return dgl.graph((neg_src, neg_dst), num_nodes=graph.num_nodes())
+
+
+def train(device, g, model):
+    opt = torch.optim.Adam(model.parameters(), lr=0.005)
+
+    for epoch in range(50):
+        model.train()
+        negative_g = construct_negative_graph(g, 1)
+        x = g.ndata["feat"].to(device)
+        pos_score, neg_score = model(g, negative_g, g, x)
+        score = torch.cat([pos_score, neg_score])
+        pos_label = torch.ones_like(pos_score).to(device)
+        neg_label = torch.zeros_like(neg_score).to(device)
+        labels = torch.cat([pos_label, neg_label])
+
+        loss = F.binary_cross_entropy_with_logits(score, labels)
+
+        opt.zero_grad()
+        loss.backward()
+        opt.step()
+
+        print(f"Epoch {epoch:05d} | Loss {loss.item():.4f}")
+
+
+def parse_args():
+    """Parses the command line arguments."""
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--mode",
+        default="cpu",
+        choices=["cpu", "gpu"],
+        help="Training mode. 'cpu' for CPU training,"
+        "'gpu' for pure-GPU training.",
+    )
+    return parser.parse_args()
+
+
+def main(args):
+    """Main function for the example."""
+    if not torch.cuda.is_available():
+        args.mode = "cpu"
+    print(f"Training in {args.mode} mode.")
+
+    # Load and preprocess dataset.
+    print("Loading data...")
+    dataset = DglLinkPropPredDataset("ogbl-citation2")
+    g = dataset[0]
+
+    g = g.to("cuda" if args.mode == "gpu" else "cpu")
+    device = torch.device("cpu" if args.mode == "cpu" else "cuda")
+    # TODO: Can we remove this in full-graph training?
```

Comment:
Are you going to address this TODO in this PR?

Reply:
Likely you just need to create reverse edges, the reverse_eids was used for filtering in sampling stage.